[KOGITO-5301][KOGITO-5444] Add Spring Boot Starters and Archetype config (#3813)
kiegroup/kie-docs: doc-content/kogito-docs/src/main/asciidoc/creating-running/chap-kogito-creating-running.adoc

--
. Open or import the project in your VSCode IDE to view the contents.

[id="proc-customizing-spring-boot-project-creation_{context}"]
=== Creating a custom {PRODUCT} Spring Boot project

> With Spring Boot, you can create custom Maven projects using {PRODUCT} Spring Boot starter or {PRODUCT} add-ons. The {PRODUCT} Spring Boot starters include:

Reviewer: Just an idea. We create a Maven project using `archetype` (and not starter) which adds necessary Spring Boot starters and add-ons. Maybe we can be more "exact" here?

--
. Open or import the project in your VSCode IDE to view the contents.

[id="proc-customizing-spring-boot-project-creation_{context}"]
=== Creating a custom {PRODUCT} Spring Boot project

> On Spring Boot, you can create custom Maven project using {PRODUCT} Spring Boot starter or {PRODUCT} add-on. The {PRODUCT} Spring Boot starters include:

Reviewer 2: ```suggestion
With Spring Boot, you can create custom Maven projects using {PRODUCT} Spring Boot starter or {PRODUCT} add-ons. The {PRODUCT} Spring Boot starters include:
```

+
--
[source,subs="attributes+"]
----
$ mvn archetype:generate \
    -DarchetypeGroupId=org.kie.kogito \
    -DarchetypeArtifactId=kogito-springboot-archetype \
    -DgroupId=org.acme -DartifactId=sample-kogito \
    -DarchetypeVersion={COMMUNITY_VERSION_FINAL} \
    -Dversion=1.0-SNAPSHOT
    -Dstarters=processes
----
After executing the command, the generated project includes the dependencies that are required to run a {PRODUCT} process services. You can combine multiple {PRODUCT} Spring Boot starters using a comma-separated list. For example `starters=processes,rules`.
--

* To generate a project containing Prometheus monitoring and Infinispan persistence using {PRODUCT} add-on, enter the following command:
+
--
[source,subs="attributes+"]
----
$ mvn archetype:generate \
    -DarchetypeGroupId=org.kie.kogito \
    -DarchetypeArtifactId=kogito-springboot-archetype \
    -DgroupId=org.acme -DartifactId=sample-kogito \
    -DarchetypeVersion={COMMUNITY_VERSION_FINAL} \
    -Dversion=1.0-SNAPSHOT
    -Dstarters=processes
    -Daddons=monitoring-prometheus,persistence-infinispan
----
////
Add a link of Kogito-add-ons section when prepared.
////

> NOTE: When you pass an add-on to the property, the add-on name does not require `kogito-addons-springboot` suffix. Also, you can combine the `add-ons` and `starters` property to customize the project.

Reviewer 2: ```suggestion
NOTE: When you pass an add-on to the property, the add-on name does not require the `kogito-addons-springboot` suffix. Also, you can combine the `add-ons` and `starters` properties to customize the project.
```

.Procedure
. In a command terminal, navigate to a local folder where you want to store the new {PRODUCT} project.

> . Enter any of the following command to generate a project using `starters` or `addons` property:

Reviewer 3: Please keep this step to better align the following bullets

Author: @hmanwani-rh we removed this line because now we have:

```
* To generate a project based on {PRODUCT} process services using the `starters` property, enter the following command:
```

Reviewer 3: The bullet points display two options for the user to generate a project. However, we need to have one single step saying that the user needs to run one of the following commands to generate a project. 
hope I'm making sense here.

Author: Oh I like that!

* {URL_PROCESS_SERVICES}#con-bpmn-variables_kogito-developing-process-services[Variables in {PRODUCT} processes]
* xref:proc-kogito-designing-app-rule-units_kogito-creating-running[]

[id="con-kogito-spring-boot-starters_{context}"]
== Spring Boot Starters for your {PRODUCT} project

The https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#using-boot-starter[Spring Boot starters] are a set of dependency descriptors that you can include in your {PRODUCT} application. Using Spring Boot starters, you can eliminate the need to go through sample code, and copy and pasting dependency descriptors.

> When you create a project, the `pom.xml` file contains all the necessary starters. You can also include a new starter if required. For more information on creating projects using Spring Boot starter, see xref:proc-customizing-spring-boot-project-creation_kogito-creating-running[].

Reviewer: Here as well, I think it should mention archetype.

=== {PRODUCT} Spring Boot Starters

The {PRODUCT} Spring Boot starter is an all-in-one descriptor for the {PRODUCT} projects that require business automation engines provided by {PRODUCT}. The {PRODUCT} Spring Boot starter includes decisions, rules, processes, predictions, and the Serverless Workflow implementation.

When your project consists of all the assets and you want a quick way to get started with {PRODUCT}, you can use the `kogito-spring-boot-starter`. For more granular approach, you can use the specific starters or a combination of starters.

You can add a {PRODUCT} Spring Boot starter to your project as follows:

[source,xml]
----
<dependencies>
  <dependency>
    <groupId>org.kie.kogito</groupId>
    <artifactId>kogito-spring-boot-starter</artifactId>
  </dependency>
</dependencies>
----

NOTE: If you plan to use Serverless Workflow with events, you can add `kogito-serverless-workflow-spring-boot-starter` to the project. The `kogito-serverless-workflow-spring-boot-starter` includes CloudEvents support required by Serverless Workflows.

The following list describes the different Spring Boot starters:

{PRODUCT} decisions Spring Boot starter::

> Starters for decisions such as DMN support for your {PRODUCT} Spring Boot project. You can add a {PRODUCT} decisions Spring Boot starter as follows:

Reviewer: ```suggestion
Starter for decisions such as DMN support for your {PRODUCT} Spring Boot project. You can add a {PRODUCT} decisions Spring Boot starter as follows:
```
AFAIK we have just one starter per asset type for now.

[source,xml]
----
<dependencies>
  <dependency>
    <groupId>org.kie.kogito</groupId>
    <artifactId>kogito-decisions-spring-boot-starter</artifactId>
  </dependency>
</dependencies>
----
{PRODUCT} predictions Spring Boot starter::

> Starters for predictions such as PMML support for your {PRODUCT} Spring Boot project. You can add a {PRODUCT} predictions Spring Boot starter as follows:

Reviewer 3: ```suggestion
Starter for predictions such as PMML support for your {PRODUCT} Spring Boot project. You can add a {PRODUCT} predictions Spring Boot starter as follows:
```

[source,xml]
----
<dependencies>
  <dependency>
    <groupId>org.kie.kogito</groupId>
    <artifactId>kogito-predictions-spring-boot-starter</artifactId>
  </dependency>
</dependencies>
----

{PRODUCT} processes Spring Boot starter::

> Starters for processes such as BPMN support for your {PRODUCT} Spring Boot project. You can add a {PRODUCT} processes Spring Boot starter as follows:

Reviewer 3: ```suggestion
Starter for processes such as BPMN support for your {PRODUCT} Spring Boot project. You can add a {PRODUCT} processes Spring Boot starter as follows:
```

[source,xml]
----
<dependencies>
  <dependency>
    <groupId>org.kie.kogito</groupId>
    <artifactId>kogito-processes-spring-boot-starter</artifactId>
  </dependency>
</dependencies>
----

{PRODUCT} rules Spring Boot starter::

> Starters for rules such as DRL support for your {PRODUCT} Spring Boot project. You can add a {PRODUCT} rules Spring Boot starter as follows:

Reviewer 3: ```suggestion
Starter for rules such as DRL support for your {PRODUCT} Spring Boot project. You can add a {PRODUCT} rules Spring Boot starter as follows:
```

[source,xml]
----
<dependencies>
  <dependency>
    <groupId>org.kie.kogito</groupId>
    <artifactId>kogito-rules-spring-boot-starter</artifactId>
  </dependency>
</dependencies>
----

{PRODUCT} Serverless Workflow Spring Boot starter::

> Starters for the Serverless Workflow support for your {PRODUCT} Spring Boot project. You can add a {PRODUCT} Serverless Workflow starter as follows:

Reviewer 3: ```suggestion
Starter for the Serverless Workflow support for your {PRODUCT} Spring Boot project. You can add a {PRODUCT} Serverless Workflow starter as follows:
```

* {URL_PROCESS_SERVICES}#con-bpmn-variables_kogito-developing-process-services[Variables in {PRODUCT} processes]
* xref:proc-kogito-designing-app-rule-units_kogito-creating-running[]

[id="con-kogito-spring-boot-starters_{context}"]
== Starters for your {PRODUCT} project

The https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#using-boot-starter[Spring Boot starters] are a set of dependency descriptors that you can include in your {PRODUCT} application. Using the Spring Boot starters, you can eliminate going through the sample code and copying-pasting various dependency descriptors.

When you create a project, the `pom.xml` file consists of all the required starters. You can also include a new starter if required. For more information about creating a project using Spring Boot starter, see xref:proc-customizing-spring-boot-project-creation_kogito-creating-running[].

[id="ref-kogito-spring-boot-starter_{context}"]
=== {PRODUCT} Spring Boot Starters

> The {PRODUCT} Spring Boot starter is an all-in-one descriptor for the {PRODUCT} projects that requires business automation engine provided by {PRODUCT}. The {PRODUCT} Spring Boot starter includes decisions, rules, processes, predictions, and the Serverless Workflow implementation.

Reviewer 2: ```suggestion
The {PRODUCT} Spring Boot starter is an all-in-one descriptor for the {PRODUCT} projects that require business automation engines provided by {PRODUCT}. The {PRODUCT} Spring Boot starter includes decisions, rules, processes, predictions, and the Serverless Workflow implementation.
```

[id="con-kogito-spring-boot-starters_{context}"]
== Starters for your {PRODUCT} project

> The https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#using-boot-starter[Spring Boot starters] are a set of dependency descriptors that you can include in your {PRODUCT} application. Using Spring Boot starters, you can eliminate the need to go through sample code, and copy & pasting dependency descriptors.

Reviewer 3: ```suggestion
The https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#using-boot-starter[Spring Boot starters] are a set of dependency descriptors that you can include in your {PRODUCT} application. Using Spring Boot starters, you can eliminate the need to go through sample code, and copy and pasting dependency descriptors.
```
